app/service/auth_svc.py

```diff
@@ -83,26 +88,48 @@ async def logout_user(request):
         raise web.HTTPFound('/login')
 
     async def login_user(self, request):
-        """
-        Log a user in and save the session
+        """Log a user in and save the session
+
         :param request:
         :return: the response/location of where the user is trying to navigate
         """
-        data = await request.post()
-        username = data.get('username')
-        password = data.get('password')
-        if self.ldap_config:
-            verified = await self._ldap_login(username, password)
-        else:
-            verified = await self._check_credentials(request.app.user_map, username, password)
-
-        if verified:
-            self.log.debug('%s logging in:' % username)
-            response = web.HTTPFound('/')
-            await remember(request, response, username)
-            raise response
-        self.log.debug('%s failed login attempt: ' % username)
-        raise web.HTTPFound('/login')
+        try:
+            self.log.debug('Using login handler "%s" for login', self._login_handler.name)
+            await self._login_handler.handle_login(request)
+        except (web.HTTPRedirection, web.HTTPUnauthorized, web.HTTPForbidden, web.HTTPSuccessful) as allowed_exception:
+            raise allowed_exception
+        except Exception as e:
+            self.log.exception('Exception when handling login request: %s', e)
+
+            # Fallback if not already using default login handler
+            if not isinstance(self._login_handler, DefaultLoginHandler):
+                self.log.debug('Falling back to default login handler')
+                return await self._default_login_handler.handle_login(request)
+            else:
+                # We ran into an unexpected exception when using the default login handler.
+                raise e
+
+    async def login_redirect(self, request, use_template=True):
+        """Redirect user to login page using the configured login handler. If using the default login handler
+        and use_template is set to true, method will return the login.html template. If use_template is set to False
+        and the default login handler is configured, it will redirect to '/login' by raising HTTPFound exception.
+        """
+
+        try:
+            self.log.debug('Using login handler "%s" for login redirect', self._login_handler.name)
+            return await self._login_handler.handle_login_redirect(request, use_template=use_template)
+        except (web.HTTPRedirection, web.HTTPUnauthorized, web.HTTPForbidden, web.HTTPSuccessful) as allowed_exception:
+            raise allowed_exception
+        except Exception as e:
+            self.log.exception('Exception when handling login redirect: %s', e)
```

Comment:
Do we not want to know the error information in case the error has more info?

Reply:
I was thinking it's unnecessary since you are using `exception()` but it could be nice to have the source exception message at the same line as the error. 

Calling `logging.exception()` includes the exception information (source exception message and stack trace) by default. The `msg` argument you pass into the `logging.exception()` call will appear before the rest of the output.

```python
import logging


def will_raise_exception():
    raise Exception("Bad request: validation failed")


def main():
    try:
        will_raise_exception()
    except:
        logging.exception("An unexpected error occurred")


if __name__ == "__main__":
    main()
```

Output:
```
ERROR:root:An unexpected error occurred  <-- the logging.exception() `msg` argument
Traceback (most recent call last):  <-- the stack trace
  File "/Users/bworrell/Desktop/exception-test.py", line 9, in main
    will_raise_exception()
  File "/Users/bworrell/Desktop/exception-test.py", line 5, in will_raise_exception
    raise Exception("Bad request: validation failed")
Exception: Bad request: validation failed  <-- the source exception message
```

Reply:
calling logging.exception will give the full stack trace with the exception info, so it should include the text from 'e'
https://docs.python.org/3/library/logging.html#logging.exception

Reply:
Gotcha, will remove the unnecessary part then